File: sw/cpp/src/bits.cpp
#include "reference/bits.hpp"

#include <stdexcept>

namespace reference {

void validate_bit_length(
    const std::span<const std::uint8_t> data,
    const std::size_t bit_length
) {
    if (bit_length > data.size() * 8U) {
        throw std::invalid_argument("bit_length exceeds data size");
    }
    const auto expected_bytes = (bit_length + 7U) / 8U;
    if (data.size() != expected_bytes) {
        throw std::invalid_argument("data does not use minimal byte length");
    }
    if (bit_length % 8U != 0U && !data.empty()) {
        const auto used_bits = static_cast<unsigned>(bit_length % 8U);
        const auto unused_mask = static_cast<std::uint8_t>(0xFFU << used_bits);
        if ((data.back() & unused_mask) != 0U) {
            throw std::invalid_argument("unused high bits must be zero");
        }
    }
}

Bits unpack_bits(
    const std::span<const std::uint8_t> data,
    const std::size_t bit_length
) {
    validate_bit_length(data, bit_length);
    Bits bits;
    bits.reserve(bit_length);
    for (std::size_t index = 0; index < bit_length; ++index) {
        bits.push_back(static_cast<std::uint8_t>(
            (data[index / 8U] >> (index % 8U)) & 1U
        ));
    }
    return bits;
}

Bytes pack_bits(const std::span<const std::uint8_t> bits) {
    Bytes output((bits.size() + 7U) / 8U, 0U);
    for (std::size_t index = 0; index < bits.size(); ++index) {
        const auto bit = bits[index];
        if (bit > 1U) {
            throw std::invalid_argument("bits must contain only 0 or 1");
        }
        output[index / 8U] |= static_cast<std::uint8_t>(bit << (index % 8U));
    }
    return output;
}

}

File: sw/common/bits.py
"""LSB-first bit packing shared by the Python golden models."""


def validate_bit_length(data: bytes, bit_length: int) -> None:
    """Validate the canonical byte representation of an LSB-first bit string.

    Parameters
    ----------
    data : bytes
        Packed bytes containing the bit string.
    bit_length : int
        Number of valid bits in ``data``.

    Raises
    ------
    ValueError
        If the length is invalid, storage is not minimal, or padding is nonzero.
    """
    if bit_length < 0 or bit_length > len(data) * 8:
        raise ValueError("bit_length must describe bits present in data")
    expected_bytes = (bit_length + 7) // 8
    if len(data) != expected_bytes:
        raise ValueError("data must use the minimal byte length for bit_length")
    if bit_length % 8 and data:
        unused_mask = (0xFF << (bit_length % 8)) & 0xFF
        if data[-1] & unused_mask:
            raise ValueError("unused high bits in the final byte must be zero")


def unpack_bits(data: bytes, bit_length: int) -> list[int]:
    """Unpack bytes in byte-zero, least-significant-bit-first order.

    Parameters
    ----------
    data : bytes
        Canonically packed bytes.
    bit_length : int
        Number of valid bits to unpack.

    Returns
    -------
    list[int]
        Bits represented as integers containing only zero or one.
    """
    validate_bit_length(data, bit_length)
    return [(data[index // 8] >> (index % 8)) & 1 for index in range(bit_length)]


def pack_bits(bits: list[int]) -> bytes:
    """Pack bits in byte-zero, least-significant-bit-first order.

    Parameters
    ----------
    bits : list[int]
        Bit values containing only zero or one.

    Returns
    -------
    bytes
        Minimally sized packed representation with zero-valued padding.

    Raises
    ------
    ValueError
        If a value is not zero or one.
    """
    output = bytearray((len(bits) + 7) // 8)
    for index, bit in enumerate(bits):
        if bit not in (0, 1):
            raise ValueError("bits must contain only 0 or 1")
        output[index // 8] |= bit << (index % 8)
    return bytes(output)


def flip_bits(data: bytes, bit_length: int, positions: list[int]) -> bytes:
    """Flip selected positions in an LSB-first packed bit string.

    Parameters
    ----------
    data : bytes
        Canonically packed bytes.
    bit_length : int
        Number of valid bits in ``data``.
    positions : list[int]
        Zero-based bit positions to invert.

    Returns
    -------
    bytes
        A copy with the requested positions inverted.

    Raises
    ------
    ValueError
        If the input or a bit position is invalid.
    """
    validate_bit_length(data, bit_length)
    output = bytearray(data)
    for position in positions:
        if position < 0 or position >= bit_length:
            raise ValueError("bit position is outside bit_length")
        output[position // 8] ^= 1 << (position % 8)
    return bytes(output)
